# Numerisk matematikk og algoritmer
## Gruppearbeid av Lars og Joakim

Vi får stort sett en feil når vi arbeider numerisk

## Definosjon av algoritme
En algoritme er en presis beskrivelse av en serie operasjoner som skal utføres for å oppnå et visst resultat.
Eksempler på domener som inneholder algoritmer:
- Datamaskiner
- Matematikk
- Sosiale strukturer og normer
- Hjernen vår

In [3]:
import math as m
# Kvadratrotalgoritme
a = 12 # Tall vi skal ta roten av
x = 3.5 # Gjetning

for i in range(3):
    x = 0.5*(x + a/x)
print("Numerisk:", x, "| Analytisk:", m.sqrt(12))

Numerisk: 3.4641016151377544 | Analytisk: 3.4641016151377544


In [4]:
import math as m
# Laget som en funksjon

def kvadrot(a,x0,N):
    """
    a: Det vi skal ta kvadratrota av
    x0: Startgjett
    N: antall iterasjoner
    """
    x = x0
    for i in range(N):
        x = 0.5*(x + a/x)
    return x

rot = kvadrot(12, 100, 10)

print("numerisk:", rot, "| analytisk:", m.sqrt(12))

numerisk: 3.4641016151377544 | analytisk: 3.4641016151377544


## Likninger og algoritmer
Her skal vi se litt på ulike måter å løse likninger på.

### Metode 1: Halveringsmetoden

In [100]:
# Halveringsmetoden - skisse
def f(x):
    return x**2 - 4

a = -10
b = 10
m = (a+b)/2

tol = 1E-15
teller = 0

while abs(f(m)) > tol:
    teller += 1
    if f(a) * f(m) < 0:
        b = m
    elif f(b) * f(m) < 0:
        a = m
    m = (a+b)/2
print(m, teller)

-2.0 54


#### Metode som funksjon

In [102]:
def g(x):
    return x**6 - 3*x**5 - 2*x**2 - 6

def halv(g, a, b, tol = 1E-14):
    """
    f : funksjon
    a : starten av interval
    b : slutten av interval
    """
    m = (a+b)/2
    teller = 0
    while abs(g(m)) > tol:
        if g(a) * g(m) < 0:
            b = m
        elif g(b) * g(m) < 0:
            a = m
        m = (a+b)/2
        teller += 1
    return m, teller

nullpunkt, antall = halv(g, -10, 10)
print("Nullpunktet er:", nullpunkt, "og løkka kjørte", antall, "ganger.")

Nullpunktet er: -1.158680400920244 og løkka kjørte 54 ganger.


### Metode 2: Newtons metode
Sjekker tangentenes nullpunkter til punkter og finner nullpunkt.
Vi har en kontinuerlig funksjon $f$. Vi tilnærmer nullpunktet til $f$ med nullpunktet til tangentene til $f(x)$, $T(x)$

Vi bruker ettpunktsformelen:

$$y - y_1 = a(x - x_1)$$

Vi om formulerer og setter $a$ til den deriverte av $x$ (siden stigningstallet kan bli funnet av den deriverte):

$$y = f(x_1) + f'(x_1)(x - x_1)$$

$$0 = f(x_1) + f'(x_1)(x - x_1)$$

$$- \frac{f(x_1)}{f'(x_1)} = x - x_1$$

$$x = x_1 - \frac{f(x_1)}{f'(x_1)}$$

Da har vi til slutt en metode som er Newtons metode:

$$x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}$$

In [104]:
def h(x):
    return x**2 - 4

def fder(x): #(f, x, dx=1E-8):
    #der = (f(x + dx) - f(x))/dx
    #return der
    #numerisk metode i kommentarer
    return 2*x

def newt(h, fder, x, tol=1E-14, N=100):
    """
    f : funksjon du vil bruke
    der : den deriverte av funksjonen
    x : gjettning
    """
    i = 0
    while abs(h(x)) > tol and i < N:
        x -= h(x)/fder(x)
        i += 1
    return x, i

nullpunkt, antall = newt(h, fder, 2000)
print("Nullpunktet er:", nullpunkt, "og løkka kjørte", antall, "ganger")

Nullpunktet er: 2.0 og løkka kjørte 15 ganger


Problemer med metoden er at den kan møte topp- og bunnpunkter som gjør at metoden ikke konvergerer.

## Skyte spurv med atombomber
### a)
Utrykket etter ei viss tid $y(t)$ kan uttrykkes ved:

$$y(t) = v_0 \cdot \sin{\theta} \cdot t - \frac{1}{2}gt^2$$

der $v_0$ er startfarten, $\sin{\theta}$ er utskytningsvinkelen, $t$ er tiden og $g$ er tyngdeakselerasjonen på 9.81$m/s^2$.

Siden man har en lengde $v_0$ og en sinus på $\sin{\theta}$ kan vi gange disse sammen for å finne høyden.
Deretter kan man gange med tiden $t$ og trekke fra tyngdeakselerasjonen siden tyngdeakselerasjonen utfører negativt arbeid på ballen i høyden.


### b)
Etter en tid $t$ kan en strekning $x(t)$ (gitt i meter) uttrykkes ved:

$$x(t) = v_0 \cdot \cos{\theta} \cdot t$$

der leddene har blitt forklart i oppgave a. ($\cos{\theta}$ er fortsatt utskytningsvinkelen).

### c) og d)



In [6]:
import pylab as p
def y(t, v0, theta):
    """
    Parameters
    ----------
    t : tid
    v0 : startfart
    theta : vinkelen til kanonen

    Returns
    -------
    TYPE
        DESCRIPTION.

    """
    return v0 * p.sin(theta) * t - 0.5*(g*t**2)

def yder(t, v0, theta):
    return v0 * p.sin(theta) - g*t

def x(t, v0, theta):
    return v0 * p.cos(theta) * t

def newtons(f, fder, x, v0, theta, tol=1E-5, N=100):
    i = 0
    while abs(f(x, v0, theta)) > tol and i <= 100:
        x -= f(x, v0, theta)/fder(x, v0, theta)
        i += 1
    return x

g = 9.81

tid = newtons(y, yder, 10, 18, p.pi/10)
avstand  = x(tid, 18, p.pi/10)
print("Tid:", tid, "| Avstand:", avstand)
print(36*p.sin(p.pi/10)/g)

Tid: 1.1340075498641349 | Avstand: 19.41309485687136
1.1340073188071464


### e)

In [18]:
import pylab as p
def y(t, v0, theta):
    """
    Parameters
    ----------
    t : tid
    v0 : startfart
    theta : vinkelen til kanonen

    Returns
    -------
    TYPE
        DESCRIPTION.

    """
    return v0 * p.sin(theta) * t - 0.5*(g*t**2)

def yder(t, v0, theta):
    return v0 * p.sin(theta) - g*t

def x(t, v0, theta):
    return v0 * p.cos(theta) * t

def newtons(f, fder, x, v0, theta, tol=1E-5, N=100):
    i = 0
    while abs(f(x, v0, theta)) > tol and i <= 100:
        x -= f(x, v0, theta)/fder(x, v0, theta)
        i += 1
    return x

g = 9.81

listeTid = []
listeVinkel = []
listeAvstand = []

for i in range(6, 2, -1):
    tid = newtons(y, yder, 10, 4.62, p.pi/i)
    avstand = x(tid, 4.62, p.pi/i)
    listeTid.append(tid)
    listeVinkel.append((p.pi/i)*180/p.pi)
    listeAvstand.append(avstand)
print("Tider:", listeTid, "\nVinkler:", listeVinkel, "\nAvstander:", listeAvstand)

Tider: [0.47095005254690075, 0.5536328500167916, 0.6660210805458796, 0.8157058851927667] 
Vinkler: [29.999999999999996, 36.0, 45.0, 59.99999999999999] 
Avstander: [1.8842887575168537, 2.0692905355021325, 2.175779863798387, 1.8842805947952914]


### f)

In [21]:
import pylab as p
def y(t, v0, theta):
    """
    Parameters
    ----------
    t : tid
    v0 : startfart
    theta : vinkelen til kanonen

    Returns
    -------
    TYPE
        DESCRIPTION.
    gir høyde på kanonball
    """
    return v0 * p.sin(theta) * t - 0.5*(g*t**2)

def yder(t, v0, theta):    
    """
    DESCRIPTION
    gir den deriverte av høyden
    """
    return v0 * p.sin(theta) - g*t

def x(t, v0, theta):
    return v0 * p.cos(theta) * t
    """
    DESCRIPTION
    gir lengde av kanonball
    """

def newtons(f, fder, x, v0, theta, tol=1E-5, N=100):
    """
    Parameters
    ----------
    f : funksjon
    fder : derverte av funksjon
    x : gjettning av tid i sekunder før ballen treffer bakken
    v0 : startfart
    theta : vinkel oppgitt i radianer
    tol : toleranse for avvik i den deriverte
    N : toleranse for antall repetisjoner av denne funksjonen

    Returns
    -------
    TYPE
        DESCRIPTION.
    gir oss en ny verdi av x som er derivert av en valgt funksjon f
    """
    i = 0
    while abs(f(x, v0, theta)) > tol and i <= 100:
        x -= f(x, v0, theta)/fder(x, v0, theta)
        i += 1
    return x

g = 9.81

tid = newtons(y, yder, 10, 4.62, p.pi/4)
avstand  = x(tid, 4.62, p.pi/4)
print("Tid:", tid, "| Avstand:", avstand)

Tid: 0.6660210805458796 | Avstand: 2.175779863798387


Resultaten vi fikk fra ett forsøk med vinkel $\frac{\pi}{4}$ var at:

Tiden før ballen traff bakken - 0.9s

Avstand - 2.52m

Meter per sekund - 4.62m/s


Vi fikk et avvik på rundt $\frac{2.18 - 2.52}{2.18} \cdot 100\% = 15,60\%$.

I forsøk er det optimalt å få under 5% avvik, som viser at vår måling er noe upresis. I tilegg fikk vi en lengre avstand enn simuleringen, som viser at vi helt klart har gjort noe galt.